### PR TITLE
Add Github Actions CI

### PR DESCRIPTION
**Pull request by jon-betts**
This also:

 * Fixes `pypy` testing
 * Separates running tests from running the coverage report
 * This means we can generate Python 2.7 and 3.9 coverage reports and merge the results
 * This allowed some `no cover` type comments to be removed
 * Adds a coverage file which enforces a minimum coverage level (not 100%)

This does not:

 * Add coverage to get higher than before
 * Add lint to the project
 * Add docs or lint to CI

### Review notes

 * This builds on: https://github.com/Pylons/pyramid_layout/pull/36
 * Actions can be seen running here: https://github.com/jon-betts/pyramid_layout/actions/runs/810254815
 * The main style of the CI script and tox was taken from https://github.com/Pylons/deform

### CODE REVIEW

**Review by digitalresistor — CHANGES_REQUESTED** [review]

File: pyramid_layout/tests/test_config.py [hunk 1]
```diff
@@ -1,8 +1,8 @@
-try: #pragma NO COVERAGE
+try:
     # python < 2.7
```
[the remark this answers is not in the document]
Comment (jon-betts) [reply pyramid_layout/tests/test_config.py:2]: Done

File: .coveragerc [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[run]
+source =
+    pyramid_layout
+
+[report]
+show_missing = True
+precision = 2
+fail_under = 98.91
```
Comment (stevepiercy) [inline .coveragerc:8]: Should this be 100?

File: tox.ini [hunk 3]
```diff
@@ -1,22 +1,27 @@
 [tox]
-envlist = 
-    py26,py27,py32,py33,pypy,cover
+envlist =
+    py27,py36,py37,py38,py39,pypy27,pypy3,coverage
```
Comment (stevepiercy) [inline tox.ini:3]: I think we might need to run coverage on both py2 and py3, then combine them, like we do for Pyramid 1.10. This assumes there are different code paths depending on Python version. I haven't checked, but we could check by running coverage manually on py2 then py3.

File: .coveragerc (hunk 2 above)
Comment (jon-betts) [reply .coveragerc:8]: It should, but that ones not on me! Seriously though this PR's focus is on getting testing going and preventing back sliding. An obvious next step would be to get this to 100.

File: tox.ini (hunk 3 above)
Comment (jon-betts) [reply tox.ini:3]: This setup effectively runs it across all versions and combines it when run through `tox`. In GHA we combine 27 and 39 with `tox -e py27,py39,coverage`.

As it goes, I think the 39 coverage is a strict superset of 27 at the moment, but of course that could change.

File: .coveragerc (hunk 2 above)
Comment (stevepiercy) [reply .coveragerc:8]: I'd say push it to 100, with the understanding that you will merge it with lower coverage and some day the coverage fairy will fix coverage.

File: tox.ini (hunk 3 above)
Comment (stevepiercy) [reply tox.ini:3]: I don't think that's true. Please have a look at https://github.com/Pylons/pyramid/blob/1.10-branch/tox.ini

File: tox.ini [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
-    python setup.py dev
-    python setup.py test
-deps =
+extras =
+    tests
+setenv =
+    COVERAGE_FILE=.coverage.{envname}
+commands =
+    python --version
+    coverage run -m pytest {posargs:}
```
Comment (bertjwregeer) [inline tox.ini:12]: Instead of running coverage directly you can also set flags for pytest to do so by default.

File: tox.ini (hunk 3 above)
Comment (bertjwregeer) [reply tox.ini:3]: @stevepiercy https://github.com/Pylons/pyramid_layout/pull/37/files#diff-03609cb60b0c6e92fb771eb8787d6722b8c31ca4c03eabc788e147acd8c6fb43R54

This does run coverage for both here, and the `tox.ini` shipped here runs coverage for all versions and environments, not just the one named py2-cover or py3-cover. That is why there is no extra testenvs necessary in this case.

File: .github/workflows/ci-tests.yml [hunk 5]
```diff
@@ -0,0 +1,78 @@
+name: Build and test
+
+on:
+    # Only on pushes to master or one of the release branches we build on push
+    push:
+        branches:
+            - master
+            - "[0-9].[0-9]+-branch"
```
Comment (stevepiercy) [inline .github/workflows/ci-tests.yml:8]: This project does not use *-branch. This line is not needed. It was intended for Pyramid's branching, so I think it just got copy-pasta'd over.

File: tox.ini (hunk 3 above)
Comment (stevepiercy) [reply tox.ini:3]: OK, I did not think to look over yonder: https://github.com/jon-betts/pyramid_layout/runs/2554929471?check_suite_focus=true#step:5:64

This is good.

**Review by stevepiercy — CHANGES_REQUESTED** [review]

File: .coveragerc (hunk 2 above)
Comment (jon-betts) [reply .coveragerc:8]: If you set this to 100 coverage will fail and so the build will fail as well

File: tox.ini (hunk 4 above)
Comment (jon-betts) [reply tox.ini:12]: I think by default `pytest` doesn't support running coverage without adding 3rd party plugins (`pytest-cov`?). I think this works pretty well without introducing any 3rd parties that aren't strictly necessary to get the job done.

File: .github/workflows/ci-tests.yml (hunk 5 above)
Comment (jon-betts) [reply .github/workflows/ci-tests.yml]: I've removed this

**Comment by jon-betts** [comment]
Hi, I've responded to, but not implemented all of the feedback given, as I feel some of it doesn't work. I'm also totally happy if you want to edit or modify this PR in anyway.

File: .coveragerc (hunk 2 above)
Comment (stevepiercy) [reply .coveragerc:8]: Oh, that's right. GHA aborts the whole thing at the first failure. I created #38 to track this. OK to leave it as is for now.

File: .github/workflows/ci-tests.yml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            - uses: actions/checkout@v2
+            - name: Setup python 3.9
+              uses: actions/setup-python@v2
+              with:
+                  python-version: 3.9
+                  architecture: x64
+
+            - run: pip install tox
+            - run: tox -e py27,py39,coverage
+#    docs:
```
Comment (stevepiercy) [inline .github/workflows/ci-tests.yml:54]: I've created a separate issue to track this. Note that Sphinx 4.0 might break a docs build. If needed, temporarily pin to <4.0, and open an issue in the relevant project.

Comment (stevepiercy) [reply .github/workflows/ci-tests.yml:54]: https://github.com/Pylons/pyramid_layout/issues/39

**Review by stevepiercy — APPROVED** [review]

**Comment by stevepiercy** [comment]
LGTM!

Ping @bertjwregeer for final review.

File: tox.ini (hunk 4 above)
Comment (bertjwregeer) [reply tox.ini:12]: Okay. We've installed pytest-cov in most of our other projects, but part of that may have also been with issues in coverage and running on Windows due to python interpreter finding issues with sub-processes.

I would prefer we use `pytest-cov` and handle things the same way in each project, but I won't let this hold up the PR.

**Comment by digitalresistor** [comment]
Thanks!

You wouldn't by chance be interested in helping maintain this project moving forward?

**Comment by jon-betts** [comment]
I don't think so sorry. This work was done under the umbrella of a project to upgrade our projects to Python >3.6 which is now coming to a close. Technically we got confirmation that this project supported the required version once I had the tests working locally, but I pushed to get some time to contribute the fixes back. I think the steer is that there's appetite for spending time on contributing fixes, but probably not on taking on maintenance.

**Comment by digitalresistor** [comment]
Alright, thanks for your contribution @jon-betts

**Pull request merged**